### PR TITLE
[spirv] Fix issues related to struct inheritance memory layout

### LINKED ISSUE #3260: [SPIR-V][Bug]: struct inheritance: generated SPIR-V is invalid
**Code**:

```hlsl
struct ProbeShL1
{
	static const uint s_coefficientCount = 4;

	float4 m_coefficients[s_coefficientCount];
};

struct Probe : ProbeShL1
{
	static const uint s_coefficientCount = ProbeShL1::s_coefficientCount;
};

RWStructuredBuffer< Probe > g_probes : register(u0);

[numthreads(64u, 1u, 1u)]
void csMain(uint3 dispatchThreadId : SV_DispatchThreadID)
{
	g_probes[dispatchThreadId.x] = (Probe)0;
}
```

**Without** -Vd:
http://shader-playground.timjones.io/d5bf806acc5f97b8891283fc31a34dda

```
fatal error: generated SPIR-V is invalid: Structure id 3 decorated as BufferBlock for variable in Uniform storage class must follow relaxed storage buffer layout rules: member 0 contains an array with stride 0
  %type_RWStructuredBuffer_Probe = OpTypeStruct %_runtimearr_Probe
```

**With** -Vd:
http://shader-playground.timjones.io/9943204e658ab2aefb43918c2a291c88

### PR DESCRIPTION
**Pull request by ehsannas**
Fixes #3260

### CODE REVIEW

File: tools/clang/test/CodeGenSPIRV/oo.inheritance.layout-differences.hlsl [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// CHECK:     [[arr0:%\d+]] = OpCompositeExtract %v4float [[arr]] 0
+// CHECK:     [[arr1:%\d+]] = OpCompositeExtract %v4float [[arr]] 1
+// CHECK:     [[arr2:%\d+]] = OpCompositeExtract %v4float [[arr]] 2
+// CHECK:     [[arr3:%\d+]] = OpCompositeExtract %v4float [[arr]] 3
+// CHECK:  [[new_arr:%\d+]] = OpCompositeConstruct %_arr_v4float_uint_4 [[arr0]] [[arr1]] [[arr2]] [[arr3]]
+// CHECK: [[new_base:%\d+]] = OpCompositeConstruct %Base [[new_arr]]
+// CHECK:        [[a:%\d+]] = OpCompositeExtract %v4float [[p]] 1
+// CHECK:    [[new_p:%\d+]] = OpCompositeConstruct %Derived [[new_base]] [[a]]
+// CHECK:                     OpStore {{%\d+}} [[new_p]]
+	g_probes[0] = p;
```
Comment (jaebaek) [inline tools/clang/test/CodeGenSPIRV/oo.inheritance.layout-differences.hlsl:38]: `p` and `g_probes[0]` have different memory layouts.
I want to doublecheck if this conversion works fine (probably no validation error means the conversion is fine).

**Review by jaebaek — COMMENTED** [review]
LGTM with a nit.

File: tools/clang/test/CodeGenSPIRV/oo.inheritance.layout-differences.hlsl (hunk 1 above)
Comment (ehsannas) [reply tools/clang/test/CodeGenSPIRV/oo.inheritance.layout-differences.hlsl:38]: Not sure if I understand what you mean. Do you mean works fine on hardware?

Comment (jaebaek) [reply tools/clang/test/CodeGenSPIRV/oo.inheritance.layout-differences.hlsl:38]: I was confused about the SPIR-V code in the `CHECK`. It shows the conversion process. Never mind.

**Review by jaebaek — APPROVED** [review]

**Pull request merged**